Further unify IO and use dask for multiple files

jni opened the pull request:
# Description

This PR fixes dask usage behaviour when reading in folders of files. It also further unifies the behaviour of the file > open dialog, the command line file IO, and the drag and drop functionality.

This PR does not:
- allow lazy loading of single files e.g. multi-page tiffs.
- work correctly with mixed dtypes or mixed shapes if the user doesn't use `use_dask=False` (because we need to assume that all dtypes are the same, or read in all files in sequence to check the dtype).
- allow the user to set `use_dask` or `stack` options with `File > open` or drag and drop.

I haven't had time to test this but it could definitely have a first round of preliminary review to check the design.

## Type of change
<!-- Please delete options that are not relevant. -->
- [ ] Bug-fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

jni committed:
Make dask and zarr requirements

jni committed:
Further unify image reading, fixing dask usage

jni committed:
Allow file dialog to select folders

Supercedes #588

jni committed:
Default to using dask only when more than one file

sofroniewn commented on napari/_qt/qt_viewer.py:
```diff
@@ -212,12 +212,14 @@ def screenshot(self):
-        self._add_files(filenames)
+        open_dialog = QFileDialog()
+        open_dialog.setWindowTitle('Select file(s) or folder(s)')
+        open_dialog.setDirectory(self._last_visited_dir)
```
you actually need to add
```python
        open_dialog.setFileMode(QFileDialog.ExistingFiles)
```
to be able to select multiple files at the same time. i have verified that works, and selecting multiple images of the same shape does the right thing

sofroniewn replied:
oh wait no - adding that blocks the ability to open the folders :-(

sofroniewn commented on napari/util/io.py:
```diff
@@ -1,66 +1,27 @@
     """Dispatch the appropriate reader given some files.
 
-    The files are assumed to all have the same type.
+    The files are assumed to all have the same shape.
```
Unfortunately when I do this I'm getting an error message due to our contrast limits calculation :-(
```python
Traceback (most recent call last):
  File "/Users/nicholassofroniew/Github/napari/napari/_qt/qt_viewer.py", line 222, in _open_images
    self._add_files(filenames)
  File "/Users/nicholassofroniew/Github/napari/napari/_qt/qt_viewer.py", line 240, in _add_files
    self.viewer.add_image(image, rgb=is_rgb(image.shape))
  File "/Users/nicholassofroniew/Github/napari/napari/components/viewer_model.py", line 470, in add_image
    visible=visible,
  File "/Users/nicholassofroniew/Github/napari/napari/layers/image/image.py", line 191, in __init__
    self._contrast_limits_range = calc_data_range(input_data)
  File "/Users/nicholassofroniew/Github/napari/napari/util/misc.py", line 282, in calc_data_range
    min_val = reduced_data.min()
  File "/anaconda3/lib/python3.7/site-packages/numpy/core/_methods.py", line 34, in _amin
    return umr_minimum(a, axis, None, out, keepdims, initial, where)
ValueError: operands could not be broadcast together with shapes (800,3) (812,3)
```

If I comment out the contrast_limit stuff it works beautifully. I know this is a problem we've been facing for a while - the offending line is this horrific one that assumes all the planes are the same shape when concatenating them
https://github.com/napari/napari/blob/e61943347401ff6fe9da669513ae86ea7ac99a19/napari/util/misc.py#L279
which is totally unnecessary as you could just take the min / max on each separately and then take the min again, the following change from line 275 on solves it
```python
        top_plane = np.asarray(data[tuple(top_plane_idx)])
        bottom_plane = np.asarray(data[tuple(bottom_plane_idx)])
        middle_plane = np.asarray(data[tuple(middle_plane_idx)])
        min_val = np.array([top_plane.min(), bottom_plane.min(), middle_plane.min()]).min()
        max_val = np.array([top_plane.max(), bottom_plane.max(), middle_plane.max()]).max()
    else:
        min_val = data.min()
        max_val = data.max()
```

sofroniewn reviewed (approved):
Looks great. Being able to add a whole folder of irregularly shapes images is pretty cool

I'm running into some problems when I try and drag-and-drop a folder or images into the viewer
```
urllib.error.URLError: <urlopen error [Errno 21] Is a directory: '/Users/nicholassofroniew/Github/image-demos/data/hymenoptera/val/ants/'>
```

also running into some problems dragging and dropping a selection of images of different sizes

```python
  File "/Users/nicholassofroniew/Github/napari/napari/_qt/qt_viewer.py", line 379, in dropEvent
    self._add_files(filenames)
  File "/Users/nicholassofroniew/Github/napari/napari/_qt/qt_viewer.py", line 240, in _add_files
    self.viewer.add_image(image, rgb=is_rgb(image.shape))
  File "/Users/nicholassofroniew/Github/napari/napari/components/viewer_model.py", line 470, in add_image
    visible=visible,
  File "/Users/nicholassofroniew/Github/napari/napari/layers/image/image.py", line 191, in __init__
    self._contrast_limits_range = calc_data_range(input_data)
  File "/Users/nicholassofroniew/Github/napari/napari/util/misc.py", line 276, in calc_data_range
    bottom_plane = np.asarray(data[tuple(bottom_plane_idx)])
  File "/anaconda3/lib/python3.7/site-packages/numpy/core/_asarray.py", line 85, in asarray
    return array(a, dtype, copy=False, order=order)
  File "/anaconda3/lib/python3.7/site-packages/dask/array/core.py", line 1047, in __array__
    x = self.compute()
  File "/anaconda3/lib/python3.7/site-packages/dask/base.py", line 156, in compute
    (result,) = compute(self, traverse=False, **kwargs)
  File "/anaconda3/lib/python3.7/site-packages/dask/base.py", line 398, in compute
    results = schedule(dsk, keys, **kwargs)
  File "/anaconda3/lib/python3.7/site-packages/dask/threaded.py", line 76, in get
    pack_exception=pack_exception, **kwargs)
  File "/anaconda3/lib/python3.7/site-packages/dask/local.py", line 462, in get_async
    raise_exception(exc, tb)
  File "/anaconda3/lib/python3.7/site-packages/dask/compatibility.py", line 112, in reraise
    raise exc
  File "/anaconda3/lib/python3.7/site-packages/dask/local.py", line 230, in execute_task
    result = _execute_task(task, data)
  File "/anaconda3/lib/python3.7/site-packages/dask/core.py", line 119, in _execute_task
    return func(*args2)
IndexError: index 533 is out of bounds for axis 1 with size 375
```

That `contrast_limits` bug needs should be fixed too.

Can we also confirm that this closes #489

It's not clear to me if we can get selecting multiple files and directories at the same time :-( i guess it should be possible, but the simple thing i did didn't seem to work

jni committed:
Revert to reading only filenames to allow multiple files

jni committed:
Add open folder command

jni committed:
Fix issue with range calculation for irregular images

jni committed:
Remove unused import

jni committed:
Add IO tests

jni committed:
Add tests for multi-tiff

jni committed:
Add test for zarr input and output

jni committed:
Fix drag and drop functionality

jni committed:
Allow single file input

jni committed:
Add fsspec to requirements

Attempting to address this failed test:
https://cirrus-ci.com/task/5454717638934528?command=test#L388

jni committed:
Add backcall to requirements

Attempting to fix:
https://cirrus-ci.com/task/6439880057421824?command=check#L132

jni committed:
Remove comments from requirements

jni committed:
Actually add folder open action to file menu

sofroniewn commented on napari/util/misc.py:
```diff
@@ -269,18 +269,19 @@ def calc_data_range(data):
 
-    min_val = reduced_data.min()
-    max_val = reduced_data.max()
+    min_val = np.min(reduced_data)
```
I think @jni this change has caused some of our tests to break when I run pytest locally - but I'm not quite sure why they went through on this PR - for example

```
napari/tests/test_numpy_like.py:31:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
napari/components/viewer_model.py:477: in add_image
    visible=visible,
napari/layers/image/image.py:191: in __init__
    self._contrast_limits_range = calc_data_range(input_data)
napari/util/misc.py:286: in calc_data_range
    min_val = np.min(reduced_data)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

args = (dask.array<random_sample, shape=(10, 15, 6, 16), dtype=float64, chunksize=(10, 15, 6, 16)>,), kwargs = {}
relevant_args = (dask.array<random_sample, shape=(10, 15, 6, 16), dtype=float64, chunksize=(10, 15, 6, 16)>, None)

>   ???
E   TypeError: no implementation found for 'numpy.amin' on types that implement __array_function__: [<class 'dask.array.core.Array'>]
```

sofroniewn replied:
ah you are well acquainted with this issue!! looks like it is fixed in the new dask release - https://github.com/dask/dask/issues/5031

Side comment I think we should add info on all our named dependencies to our about menu! it's just so convenient

sofroniewn replied:
all tests now pass for me
